### PR TITLE
2599 task add an istemporary flag to all assets

### LINKED ISSUE #2599: [TASK]: Add an `isTemporary` flag to all assets
**Describe the task**

A temporary asset is an asset that needs to exist in the data-service for processing task within operators in the workflow.
But they should never appears in search, nor in project. They are only visible in the hmi-client to the operator that manipulate them.
We need a simple way to define them.

**Task Items**
- [x] add a way for the `hmi-client` to create _temporary_ assets
- [x] those _temporary_ assets should not be searchable, nor visible within a `Project.projectAssets[]`
- [x] a _temporary_ asset can become a normal asset easily
- [x] a _normal_ asset cannot become a _temporary_ asset
- [x] Needs to be available on all assets except _Project_

### PR DESCRIPTION
**Pull request by dvince2**
# Description

A few things in motion here:
* The `temporary` flag is set via the standard `put` endpoint to update an asset
* If you try to go from permanent to temporary an exception is thrown
* Some major refactoring and cleanup here, moving most of the asset services into a single class and using generics. This will reduce the amount of duplicate code a lot.
* In the controller classes I have also added a bunch of safety code around NPEs and empty `Optional`s.

Resolves #2599

### CODE REVIEW

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/ArtifactController.java [hunk 1]
```diff
@@ -135,17 +121,14 @@ public ResponseEntity<Artifact> getArtifact(@PathVariable("id") UUID artifactId)
...
 		try {
 			artifact.setId(artifactId);
-			Optional<Artifact> updated = artifactService.updateArtifact(artifact);
-			if (updated.isEmpty()) {
-				return ResponseEntity.notFound().build();
-			}
-			return ResponseEntity.ok(updated.get());
-		} catch (Exception e) {
+			final Optional<Artifact> updated = artifactService.updateAsset(artifact);
+			return updated.map(ResponseEntity::ok).orElseGet(() -> ResponseEntity.notFound().build());
```
Comment (dgauldie) [inline packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/ArtifactController.java:130]: `notFound` returns a 404... you will need to update the ApiResponse

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/ArtifactController.java [hunk 2]
```diff
@@ -210,12 +193,9 @@ public ResponseEntity<PresignedURL> getDownloadURL(
...
 
 		try {
-			Optional<PresignedURL> url = artifactService.getDownloadUrl(id, filename);
-			if (url.isEmpty()) {
-				return ResponseEntity.notFound().build();
-			}
-			return ResponseEntity.ok(url.get());
-		} catch (Exception e) {
+			final Optional<PresignedURL> url = artifactService.getDownloadUrl(id, filename);
+			return url.map(ResponseEntity::ok).orElseGet(() -> ResponseEntity.notFound().build());
```
Comment (dgauldie) [inline packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/ArtifactController.java:197]: `notFound` returns a 404... you will need to update the ApiResponse

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/DatasetController.java [hunk 3]
```diff
@@ -292,11 +266,8 @@ public ResponseEntity<PresignedURL> getDownloadURL(
 			@RequestParam("filename") final String filename) {
 
 		try {
-			Optional<PresignedURL> url = datasetService.getDownloadUrl(id, filename);
-			if (url.isEmpty()) {
-				return ResponseEntity.notFound().build();
-			}
-			return ResponseEntity.ok(url.get());
+			final Optional<PresignedURL> url = datasetService.getDownloadUrl(id, filename);
+			return url.map(ResponseEntity::ok).orElseGet(() -> ResponseEntity.notFound().build());
```
Comment (dgauldie) [inline packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/DatasetController.java:270]: `notFound` returns a 404... you will need to update the ApiResponse

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/DocumentController.java [hunk 4]
```diff
@@ -407,8 +415,12 @@ public ResponseEntity<AddDocumentAssetFromXDDResponse> createDocumentFromXDD(
 			// get preliminary info to build document asset
 			final Document document = body.getDocument();
 			final UUID projectId = body.getProjectId();
-			final String doi = documentAssetService.getDocumentDoi(document);
-			final String userId = projectService.getProject(projectId).get().getUserId();
+			final String doi = DocumentAsset.getDocumentDoi(document);
+			final Optional<Project> project = projectService.getProject(projectId);
+			if(project.isEmpty()) {
+				return ResponseEntity.notFound().build();
```
Comment (dgauldie) [inline packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/DocumentController.java:421]: notFound returns a 404... you will need to update the ApiResponse

**Review by dgauldie — CHANGES_REQUESTED** [review]

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/ArtifactController.java (hunk 1 above)
Comment (dvince2) [reply packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/ArtifactController.java:131]: fixed

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/ArtifactController.java (hunk 2 above)
Comment (dvince2) [reply packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/ArtifactController.java:199]: fixed

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/DatasetController.java [hunk 5]
```diff
@@ -175,17 +152,14 @@ public ResponseEntity<ResponseDeleted> deleteDataset(
...
 		try {
 			dataset.setId(id);
-			final Optional<Dataset> updated = datasetService.updateDataset(dataset);
-			if (updated.isEmpty()) {
-				return ResponseEntity.notFound().build();
-			}
-			return ResponseEntity.ok(updated.get());
-		} catch (IOException e) {
+			final Optional<Dataset> updated = datasetService.updateAsset(dataset);
+			return updated.map(ResponseEntity::ok).orElseGet(() -> ResponseEntity.notFound().build());
```
[the remark this answers is not in the document]
Comment (dvince2) [reply packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/DatasetController.java:162]: fixed

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/DatasetController.java [hunk 6]
```diff
@@ -504,11 +491,11 @@ private ResponseEntity<ResponseStatus> uploadCSVAndUpdateColumns(final UUID data
...
 
 	private static List<List<String>> csvToRecords(final String rawCsvString) throws IOException {
-		List<List<String>> records = new ArrayList<>();
-		try (CSVParser parser = new CSVParser(new StringReader(rawCsvString), CSVFormat.DEFAULT)) {
-			for (CSVRecord record : parser) {
-				List<String> values = new ArrayList<>();
-				record.forEach(cell -> values.add(cell));
+		final List<List<String>> records = new ArrayList<>();
+		try (final CSVParser parser = new CSVParser(new StringReader(rawCsvString), CSVFormat.DEFAULT)) {
+			for (final CSVRecord record : parser) {
```
[the remark this answers is not in the document]
Comment (dvince2) [reply packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/DatasetController.java:498]: fixed

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/DatasetController.java (hunk 3 above)
Comment (dvince2) [reply packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/DatasetController.java:272]: fixed

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/ModelController.java [hunk 7]
```diff
@@ -140,19 +123,19 @@ ResponseEntity<Model> getModel(@PathVariable("id") UUID id) {
...
 						// Fetch the Document extractions
 						final Optional<DocumentAsset> document = documentAssetService
-								.getDocumentAsset(UUID.fromString(documentId));
+								.getAsset(UUID.fromString(documentId));
 						if (document.isPresent()) {
 							final List<JsonNode> extractions = objectMapper.convertValue(
 									document.get().getMetadata().get("attributes"),
-									new TypeReference<List<JsonNode>>() {
-									});
+                                    new TypeReference<>() {
```
Comment (dgauldie) [inline packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/ModelController.java:130]: what is going on with the tabbing??!?!?!

**Review by dgauldie — APPROVED** [review]

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/ModelController.java (hunk 7 above)
Comment (dvince2) [reply packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/ModelController.java]: woah. Fixed!

**Review by YohannParis — APPROVED** [review]
LGTM.
Why not take advantage of that huge tiding up and remove Artifacts!

**Pull request merged**